### PR TITLE
Add urlCluster table function and refactor all *Cluster table functions

### LINKED ISSUE #38499: `urlCluster` table function
**Use case**

Distributed processing of a large number of files.

**Describe the solution you'd like**

Similar to `s3Cluster` and `hdfsCluster`

**Additional context**

Note that the url table function only supports `{...}` expansions, not `*` or `?`.

### PR DESCRIPTION
**Pull request by attack204**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add  urlCluster table function. Refactor all *Cluster table functions to reduce code duplication. Make schema inference work for all possible *Cluster function signatures and for named collections.
Closes https://github.com/ClickHouse/ClickHouse/issues/38499

### CODE REVIEW

File: src/Storages/StorageURLCluster.cpp [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    if (columns_.empty())
+    {
+        auto columns = StorageURL::getTableStructureFromData(format_name_,
+            uri_,
+            chooseCompressionMethod(Poco::URI(uri_).getPath(), compression_method),
+            configuration_.headers,
+            std::nullopt,
+            context_);
+        storage_metadata.setColumns(columns);
+        add_columns_structure_to_query = true;
```
Comment (Avogar) [inline src/Storages/StorageURLCluster.cpp:64]: I see you almost copied it from `StorageS3Cluster`. It's ok, but please take into account recent fix https://github.com/ClickHouse/ClickHouse/pull/45422

Comment (attack204) [reply src/Storages/StorageURLCluster.cpp:64]: I will fix it recently, actually I copied code from `StorageHDFSCluster`, so maybe the same bug in StorageHDFSCluter

Comment (Avogar) [reply src/Storages/StorageURLCluster.cpp:64]: Yes, there is the same bug there. I will fix it. Thanks for it :)

Comment (attack204) [reply src/Storages/StorageURLCluster.cpp]: I have doubts about the compatibility of urlCluster and RemoteQueryExecutor, in RemoteQueryExecutor, the TaskIterator only support String
```cpp
using TaskIterator = std::function<String()>;
```
The input of the S3 and HDFS' `initialize` function is a String, so they are compatible.

```cpp
StorageS3Source::ReaderHolder StorageS3Source::createReader()
{
    auto [current_key, info] = (*file_iterator)(); //current_key is a String
    if (current_key.empty())
        return {};
   ....
}
```
```cpp
bool HDFSSource::initialize()
{
    current_path = (*file_iterator)(); //current_path is a String
    if (current_path.empty())
        return false;
    ....
}
```

However, the input of the URL `initialize` function is `vector<String>`.

```cpp
    initialize = [=, this](const URIInfo::FailoverOptions & uri_options) //FailoverOptions is a std::vector<String> 
    {
        if (uri_options.empty())
            throw Exception(ErrorCodes::LOGICAL_ERROR, "Got empty url list");

        auto first_option = uri_options.begin();
    }
```

So the question is how can I pass a std::vector<String> parameter by TaskIterator?

I think there are two ways

1. provide another TaskIterator with type `std::vector<String>`
2. use some trick method to transfer std::vector<String> to String and split it in the cluster.

Comment (ucasfl) [reply src/Storages/StorageURLCluster.cpp]: The first one is better.

**Comment by alexey-milovidov** [comment]
Let's fix the tests and continue.

**Comment by attack204** [comment]
> Let's fix the tests and continue.

I'm very sorry, I may not be able to participate in the open-source community for a long time(maybe more than 4 months). I am looking for job opportunities now, but the recruitment situation in China this year is very challenging(most of the undergraduates have no job), so I have to spend the most time preparing for the job interview.

**Comment by alexey-milovidov** [comment]
@attack204, No problem, we can continue when you will have time.

By the way, you are already officially, a ClickHouse contributor, because you have contributed a valuable change:
https://github.com/ClickHouse/ClickHouse/pull/46161

Moreover, your name is present in every new ClickHouse release:
```
milovidov-desktop :) SELECT * FROM system.contributors WHERE name LIKE '%attack%'

SELECT *
FROM system.contributors
WHERE name LIKE '%attack%'

Query id: d321d57d-9138-40c9-a1dc-e499e6e18169

┌─name──────┐
│ attack204 │
└───────────┘

1 row in set. Elapsed: 0.001 sec.
```

I hope it can help with interviews :)

**Comment by Avogar** [comment]
@attack204 thank you again for the contribution! This is really nice feature, I will finish this PR by myself

**Review by Avogar — APPROVED** [review]

**Comment by robot-ch-test-poll2** [comment]
This is an automated comment for commit 70a8fd2c50ecaad74ab41d371bac76a9f1da74b2 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/45427/70a8fd2c50ecaad74ab41d371bac76a9f1da74b2/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>🟡 pending</td></tr>
</table>

File: tests/queries/0_stateless/02414_all_new_table_functions_must_be_documented.sql [hunk 2]
```diff
@@ -2,5 +2,5 @@
 -- Please help shorten this list down to zero elements.
 SELECT name FROM system.table_functions WHERE length(description) < 10
 AND name NOT IN (
-    'cosn', 'oss', 'hdfs', 'hdfsCluster', 'hive', 'mysql', 'postgresql', 's3', 's3Cluster', 'sqlite' -- these functions are not enabled in fast test
+    'cosn', 'oss', 'hdfs', 'hdfsCluster', 'hive', 'mysql', 'postgresql', 's3', 's3Cluster', 'sqlite', 'urlCluster' -- these functions are not enabled in fast test
```
Comment (alexey-milovidov) [inline tests/queries/0_stateless/02414_all_new_table_functions_must_be_documented.sql:5]: This is wrong.

**Pull request merged**